Add moosetools submodule

aeslaughter opened the pull request:
(refs #17006)

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

aeslaughter committed:
Add moosetools submodule

(refs #17006)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/690162/) on 9a1e7c5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17108/site)

This comment will be updated on new commits.

Commit:
Replaced makefile with moosetools HIT, removed old HIT unit tests

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 0def3919634c1991d65b47edf3e330b9ca7257f0
+Subproject commit 79d440f4767c1500c766276f2df9235ea75b8fca
```
**Caution!** This contains a submodule update

aeslaughter replied:
This probably wasn't intentional.

aeslaughter commented on unit/src/HitTests.C:
```diff
@@ -1,626 +0,0 @@
-       "foo",
-       "42",
-       hit::Field::Kind::String},
-      {"fparse", "foo=${fparse 40 + 2}\n", "foo", "42", hit::Field::Kind::Float},
```
Any test with fparse needs to stay, this feature is not tested in the moosetools repository because fparser is not included. All the others can be removed.

aeslaughter commented on __init__.py:
```diff
@@ -0,0 +1,8 @@
+#* This file is part of MOOSETOOLS repository
```
This should not be needed.

aeslaughter reviewed (commented):
You can also remove framework/contrib/hit

Commit:
Fixing up merge issues

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 0def3919634c1991d65b47edf3e330b9ca7257f0
+Subproject commit 79d440f4767c1500c766276f2df9235ea75b8fca
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

permcody committed:
Fixup build issues.

hit_DIR should be defined in the included support file, not the main level file. Otherwise
we won't pick it up when building anything other than the framework (e.g. test or modules).

refs #17108

aeslaughter commented:
This has submodule update for large_media and petsc, that doesn't seem necessary.

aeslaughter reviewed (commented):
What happens if the moosetools submodule has not been inited? Do we need the Makefile to check and error or just do it for you?

permcody commented on framework/moose.mk:
```diff
@@ -38,7 +38,7 @@ pcre_deps      := $(patsubst %.cc, %.$(obj-suffix).d, $(pcre_srcfiles)) \
 # hit (new getpot parser)
 #
-hit_DIR       := $(FRAMEWORK_DIR)/contrib/hit
+hit_DIR       ?= $(MOOSE_DIR)/moosetools/contrib/hit
```
Yes, I agree that this needs to be renamed. The build problems are now gone. It appears that the unit tests are now broken though.

permcody replied:
Yes, I still need to check for the submodule not being inited.

aeslaughter replied:
That is probably just too much being removed. All of the unit tests have been moved to moosetools, except for the use of fparser. I didn't include the fparser in moosetools, so those tests need to stay in place.

permcody replied:
The issue here is that you've deleted or moved some of the types. "gtest" probably shouldn't move because we need a way to unit test MOOSE. However, I do see the dilemma of only having "gtest" in MOOSE. That would mean we can't unit test the HIT parser...

permcody replied:
I'm not sure what the goal is here. How would you expect MOOSE to work without HIT (in the moosetools repo)? I don't see how we can make this a soft dependency. If it's not a soft dependency, then we probably don't need to cleanup the unit tests (assume that gtest and HIT are always available).

permcody committed:
Fixup unit tests (missing structs)

permcody committed:
Change case of hit_DIR to be consistent with other environment variables

permcody committed:
Make "moosetools" optional - sortof. HIT is still required

refs #17108

moosebuild commented on petsc:
```diff
@@ -1 +1 @@
-Subproject commit 94fb59f4ccbca51eac1f1822a93eb939f0d846b6
+Subproject commit 747052ed59d6880432deb92f2e15b213e649f5f3
```
**Caution!** This contains a submodule update

aeslaughter commented on framework/moose.mk:
```diff
@@ -38,17 +38,21 @@ pcre_deps      := $(patsubst %.cc, %.$(obj-suffix).d, $(pcre_srcfiles)) \
-hit_DIR       := $(FRAMEWORK_DIR)/contrib/hit
-hit_srcfiles  := $(hit_DIR)/parse.cc $(hit_DIR)/lex.cc $(hit_DIR)/braceexpr.cc
+HIT_DIR       ?= $(MOOSE_DIR)/moosetools/contrib/hit
+HIT_CONTENT   := $(shell ls $(HIT_DIR) 2> /dev/null)
+ifeq ($(HIT_CONTENT),)
+  $(error The HIT parser does not seem to be available. Make sure that either the moosetools submodule is checked out or that your HIT_DIR points to the correct location of your HIT parser)
+endif
```
```suggestion
hit_CONTENT   := $(shell ls $(HIT_DIR) 2> /dev/null)
ifeq ($(hit_CONTENT),)
  $(error The HIT input file parser does not seem to be available. Make sure that either the moosetools submodule is checked out or that your HIT_DIR environment variable is set to the correct location of your HIT parser. To get the moosetools submodule, run the following from the root directory of MOOSE.\n    git submodule update --init moosetools)
endif
```

aeslaughter reviewed (commented):
One minor suggestion, you can or leave the changes.

aeslaughter commented:
@milljm @loganharbour @permcody Given that this is passing, I am guessing that all our recipes do a `git submodule update`, or am I missing something where this is happening in the Makefile?

If not in the Makefile, we will need to update our manual install instructions. We could also just add that to "update_and_rebuild_libmesh.sh" and/or make a new script "update_and_rebuild_<hit/moosetools?>.sh".

Finally, do we need to change anything in spack or conda scripts before we pull the trigger on this?

loganharbour commented:
> @milljm @loganharbour @permcody Given that this is passing, I am guessing that all our recipes do a `git submodule update`, or am I missing something where this is happening in the Makefile?

The top level submodules do get initialized and updated both with moose as the primary repo and for applications that have the moose submodule, so I believe we're good on this front.

> If not in the Makefile, we will need to update our manual install instructions. We could also just add that to "update_and_rebuild_libmesh.sh" and/or make a new script "update_and_rebuild_<hit/moosetools?>.sh".
> 
> Finally, do we need to change anything in spack or conda scripts before we pull the trigger on this?

This is the part that I'm looking forward to the least. We need to figure out if we want _all_ users to init/update moosetools, or if we want only manual users to update it and for conda users to just have an additional package.

aeslaughter commented:
I think a moose-hit package might be the way to go. As we go forward, I don't want the moosetools repository required to compile an application and run an input file.

permcody commented:
@aeslaughter - The other common package between MOOSE and the MOOSETools repo is gtest. We can't build MOOSEUnit without that. Are you thinking of making unit give a useful warning if it can't find gtest or what are you thoughts here? Maybe we don't need gtest in the moose tools repo and we leave it with MOOSE?

permcody committed:
Update framework/moose.mk with better missing HIT message.

Co-authored-by: Andrew E Slaughter <slaughter98@gmail.com>

aeslaughter commented:
There is a copy of gtest in moose and moosetools

aeslaughter committed:
Revert this to remove HIT...again

aeslaughter committed:
Update Makefile to support framework or moosetools HIT

(refs #17006)

aeslaughter committed:
Update moosetools submodule
(refs #17006)

aeslaughter committed:
Use udate_and_rebuild_hit in Makefile error

(refs #17006)

aeslaughter commented:
This includes the submodule and minimum changes so that if moosetools exists, then use it for hit, if not use framework/contrib/hit

aeslaughter commented:
@loganharbour @milljm

loganharbour commented on scripts/update_and_rebuild_hit.sh:
```diff
@@ -0,0 +1,44 @@
+
+cd $SCRIPT_DIR/..
+
+# Test for git repository when not using fast
```
```suggestion
```

loganharbour commented on scripts/update_and_rebuild_hit.sh:
```diff
@@ -0,0 +1,44 @@
+  cd $SCRIPT_DIR/..
+  echo "Usage: $0 [-h | --help | --skip-submodule-update ]"
+  echo
+  echo "-h | --help              Display this message and list of available PETSc options"
```
```suggestion
  echo "-h | --help              Display this message"
```

loganharbour commented on python/pyhit/__init__.py:
```diff
@@ -1,13 +1,27 @@
+moose_dir = mooseutils.git_root_dir(os.path.dirname(__file__))
+status = mooseutils.git_submodule_status(moose_dir)
+
+# Use framework/contrib/hit
```
```suggestion
# Use framework/contrib/hit because moosetools submodule is not available
```

loganharbour commented on python/pyhit/__init__.py:
```diff
@@ -1,13 +1,27 @@
+        moose_test_dir = os.path.abspath(os.path.join(moose_dir, 'test'))
+        subprocess.run(['make', 'hit'], cwd=moose_test_dir)
+        from . import hit
+else:
```
```suggestion
# Use hit in moosetools submodule
else:
```

loganharbour commented on framework/moose.mk:
```diff
@@ -38,17 +38,25 @@ pcre_deps      := $(patsubst %.cc, %.$(obj-suffix).d, $(pcre_srcfiles)) \
+$(info Using HIT from $(HIT_DIR))
+hit_CONTENT   := $(shell ls $(HIT_DIR) 2> /dev/null)
+ifeq ($(hit_CONTENT),)
+  $(error The HIT input file parser does not seem to be available. Make sure that either the moosetools submodule is checked out or that your HIT_DIR environment variable is set to the correct location of your HIT parser. To get the moosetools submodule, run the following from the root directory of MOOSE.\n    scripts/update_and_rebuild_hit.sh)
```
```suggestion
  $(error The HIT input file parser does not seem to be available. Make sure that either the moosetools submodule is checked out or that your HIT_DIR environment variable is set to the correct location of your HIT parser. To get the moosetools submodule, run the following from the root directory of MOOSE.\n\n    scripts/update_and_rebuild_hit.sh)
```

We should never hit this for now, right?

aeslaughter replied:
Correct. This would be if you set HIT_DIR to something that does not exist.

aeslaughter committed:
Apply suggestions from code review

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

permcody reviewed (approved):
Looks good to me!

permcody commented on .gitmodules:
```diff
@@ -8,3 +8,6 @@
     url = https://gitlab.com/petsc/petsc.git
+[submodule "moosetools"]
+	path = moosetools
+	url = git@github.com:idaholab/moosetools
```
Oh, I take that back. This should be relative path. This makes it more versatile in that it'll work with HTTPS or SSH. Let's change that before merging!

permcody replied:
I'm not even sure I know what that should be, you'll have to play with it to make sure we get the right relative path. I'd try `url = ../moosetools` as a first attempt and see if it registers properly when you init.

aeslaughter replied:
Didn't know this was a thing, it should be correct now. It is the same as large_media.

aeslaughter committed:
Make gitmodules use relative URL

(refs #17006)
